sdk/core/Azure.Core/src/DynamicData/DynamicData.AllowList.cs

```diff
@@ -0,0 +1,217 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text.Json;
+using Azure.Core.Json;
+
+namespace Azure.Core.Dynamic
+{
+    public partial class DynamicData
+    {
+        internal class AllowList
+        {
+            public static void AssertAllowedType<T>(T value)
+            {
+                if (value == null)
+                {
+                    return;
+                }
+
+                Type type = value.GetType();
+
+                if (!IsAllowedType(type, value))
+                {
+                    throw new NotSupportedException($"Type is not currently supported: '{type}'.");
+                }
+            }
+
+            public static bool IsAllowedType<T>(Type type, T value)
+            {
+                if (IsAllowedLeafType(type))
+                {
+                    return true;
+                }
+
+                if (IsAllowedCollectionType(type, value))
+                {
+                    return true;
+                }
+
+                return IsAllowedAnonymousType(type, value, new HashSet<Type>());
+            }
+
+            private static bool IsAllowedLeafType(Type type)
+            {
+                return IsAllowedPrimitive(type) ||
+                    type == typeof(string) ||
+                    type == typeof(DateTime) ||
+                    type == typeof(DateTimeOffset) ||
+                    type == typeof(TimeSpan) ||
+                    type == typeof(Uri) ||
+                    type == typeof(Guid) ||
+                    type == typeof(ETag) ||
+
+                    // Allowable for DynamicData assignments.
+                    type == typeof(JsonElement) ||
+                    type == typeof(JsonDocument) ||
+                    type == typeof(MutableJsonDocument) ||
+                    type == typeof(MutableJsonElement) ||
+                    type == typeof(DynamicData);
+            }
+
+            private static bool IsAllowedPrimitive(Type type)
+            {
+                return
+                    type.IsPrimitive ||
+                    type == typeof(decimal);
```

Comment:
Why is decimal here and not in the method above, i.e. how is decimal different from let's say ETag or DateTime?

Reply:
I had thought decimal was a primitive like double that wasn't added to `IsPrimitive` in order not to change its behavior.  But if that's incorrect, I'm happy to move it out, I can do some research to verify my claim.

Reply:
I am not sure why it matters whether decimal is primitive or not. String is also a primitive.

Reply:
I can move decimal out of this method -- if it is removed, there isn't a reason to have a separate method anymore.

Reply:
Refactored per feedback.